### PR TITLE
proxy: prevent connection leakage

### PR DESCRIPTION
**Pull request by tw4452852**
Signed-off-by: Tw <tw19881113@gmail.com>

### CODE REVIEW

**Comment by tw4452852** [comment]
Hold on, test may failed sometimes, because it closes the connection before writing the data...
Need work out a better solution.

**Comment by nemothekid** [comment]
Not sure if there is an actual connection leak here, or if capturing the error actually fixes that leak.

**Comment by mholt** [comment]
Maybe the error could be logged?

But good point, .Close() does seem to be called either way, I think, although I haven't confirmed there *isn't* a leak.

File: middleware/proxy/reverseproxy.go [hunk 1]
```diff
@@ -228,10 +228,16 @@ func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request, extr
...
-			io.Copy(backendConn, conn) // write tcp stream to backend.
+			_, err := io.Copy(backendConn, conn) // write tcp stream to backend.
+			errCh <- err
 		}()
-		io.Copy(conn, backendConn) // read tcp stream from backend.
+		go func() {
+			_, err := io.Copy(conn, backendConn) // read tcp stream from backend.
+			errCh <- err
+		}()
+		<-errCh
```
Comment (weingart) [inline middleware/proxy/reverseproxy.go]: This looks like you're not collecting everything from the error channel.

**Comment by tw4452852** [comment]
@weingart @mholt @nemothekid  Yes, we should log the error here and there is also another problem here: A websocket client connects to our proxy and our proxy makes another connection to the real server. If client close its connection, we will leave a connection(proxy -> server) unclosed and goroutine
running. This is the leakage what I mean and I want to fix it. Any ideas?

**Comment by weingart** [comment]
You are writing to the channel 2 times, and reading once.

**Comment by nemothekid** [comment]
@tw4452852 Do you see this problem in practice? I can't see how this happens in practice. There are two things that happens

1. The client closes cleanly and sends a CLOSE frame. The backend server receives this CLOSE and should close the connection. From the [Websocket RFC 6455](https://tools.ietf.org/html/rfc6455#page-36) ```After both sending and receiving a Close message, an endpoint
   considers the WebSocket connection closed and MUST close the
   underlying TCP connection.```
If your server isn't doing this, and leaving that connection open, I would check the implementation. In any case, this will lead to 2.

2.  The client doesn't close cleanly (it dies) or the server ignores it's close request. In this case, the backend server should try to send PING requests/PONG requests (or any data really). Because the client is dead, the `io.Copy` on [line 234](https://github.com/mholt/caddy/blob/master/middleware/proxy/reverseproxy.go#L234) will end, and the function will return. When this function returns, both `conn` and `backendConn` are [closed](https://github.com/mholt/caddy/blob/master/middleware/proxy/reverseproxy.go#L221) and so shouldn't be leaked. When this function returns, and the connections are closed, the `Copy` should error out and the goroutine should end.

**Comment by tw4452852** [comment]
@nemothekid Yes, here is my environment:

Client:
```go
package main

import (
        "log"
        "time"

        "github.com/gorilla/websocket"
)

func main() {
        c, _, err := websocket.DefaultDialer.Dial("ws://127.0.0.1:2020/echo", nil)
        if err != nil {
                log.Fatal("dial:", err)
        }
        defer c.Close()

        go func() {
                defer c.Close()
                for {
                        _, message, err := c.ReadMessage()
                        if err != nil {
                                log.Println("read:", err)
                                break
                        }
                        log.Printf("recv: %s", message)
                }
        }()

        ticker := time.NewTicker(time.Second)
        defer ticker.Stop()

        for t := range ticker.C {
                log.Printf("send: %s\n", t)
                err := c.WriteMessage(websocket.TextMessage, []byte(t.String()))
                if err != nil {
                        log.Println("write:", err)
                        break
                }
        }
}
```

Server:
```go
package main

import (
        "flag"
        "log"
        "net/http"

        "github.com/gorilla/websocket"
)

var addr = flag.String("addr", "127.0.0.1:2021", "http service address")

var upgrader = websocket.Upgrader{} // use default options

func echo(w http.ResponseWriter, r *http.Request) {

        c, err := upgrader.Upgrade(w, r, nil)
        if err != nil {
                log.Print("upgrade:", err)
                return
        }
        defer c.Close()
        for {
                mt, message, err := c.ReadMessage()
                if err != nil {
                        log.Println("read:", err)
                        break
                }
                log.Printf("recv: %s", message)
                err = c.WriteMessage(mt, message)
                if err != nil {
                        log.Println("write:", err)
                        break
                }
        }
}

func main() {
        flag.Parse()
        http.HandleFunc("/echo", echo)
        log.Fatal(http.ListenAndServe(*addr, nil))
}
```

Caddyfile
```
127.0.0.1:2020 {
  log proxy.log
   proxy / 127.0.0.1:2021 {
        proxy_header Host {host}
        proxy_header X-Real-IP {remote}
        proxy_header X-Forwarded-For {remote}
        websocket
    }
}
```

Start them all and then press `ctrl-c` to stop the client, you will see a goroutine wait on the connection:

```
goroutine 8 [IO wait]:
net.runtime_pollWait(0x7fbc01c2cda8, 0x72, 0xc8201ba000)
        /home/tw/goroot/src/runtime/netpoll.go:160 +0x4f
net.(*pollDesc).wait(0xc820122fb0, 0x72, 0xc8200d90b8, 0xc82000c0f8)
        /home/tw/goroot/src/net/fd_poll_runtime.go:73 +0x2e
net.(*pollDesc).waitRead(0xc820122fb0, 0xbce6c0, 0xc82000c0f8)
        /home/tw/goroot/src/net/fd_poll_runtime.go:78 +0x2a
net.(*netFD).Read(0xc820122f50, 0xc8201ba000, 0x8000, 0x8000, 0x0, 0xbce6c0, 0xc82000c0f8)
        /home/tw/goroot/src/net/fd_unix.go:214 +0x183
net.(*conn).Read(0xc8200300c8, 0xc8201ba000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
        /home/tw/goroot/src/net/net.go:172 +0x63
io.copyBuffer(0x7fbc01c31160, 0xc8201321a0, 0xbcbfc0, 0xc8200300c8, 0xc8201ba000, 0x8000, 0x8000, 0x9303e0, 0xc8200d9200, 0xbcbfc0)
        /home/tw/goroot/src/io/io.go:397 +0x13a
io.Copy(0x7fbc01c31160, 0xc8201321a0, 0xbcbfc0, 0xc8200300c8, 0xc8200300c8, 0xc8201321a0, 0x0)
        /home/tw/goroot/src/io/io.go:367 +0x5a
github.com/mholt/caddy/middleware/proxy.(*ReverseProxy).ServeHTTP(0xc8201703c0, 0xbd0f40, 0xc82001ecd0, 0xc8201aa000, 0xc8200d94b0, 0x0, 0x0)
        /home/tw/golib/src/github.com/mholt/caddy/middleware/proxy/reverseproxy.go:234 +0x837
github.com/mholt/caddy/middleware/proxy.Proxy.ServeHTTP(0xbcb380, 0xc820164600, 0xc82016e820, 0x1, 0x1, 0xbd0f40, 0xc82001ecd0, 0xc8201aa000, 0x8000104, 0x0, ...)
        /home/tw/golib/src/github.com/mholt/caddy/middleware/proxy/proxy.go:125 +0x40c
github.com/mholt/caddy/middleware/proxy.(*Proxy).ServeHTTP(0xc820164630, 0xbd0f40, 0xc82001ecd0, 0xc8201aa000, 0xbcf3c0, 0xc82011ac60, 0x9bb448)
        <autogenerated>:5 +0x7a
github.com/mholt/caddy/middleware/log.Logger.ServeHTTP(0xbcd880, 0xc820164630, 0xc820166780, 0x1, 0x1, 0x9bdd50, 0xbd1200, 0xc8200cadd0, 0xc8201aa000, 0xc820168750, ...)
        /home/tw/golib/src/github.com/mholt/caddy/middleware/log/log.go:31 +0x232
github.com/mholt/caddy/middleware/log.(*Logger).ServeHTTP(0xc820164660, 0xbd1200, 0xc8200cadd0, 0xc8201aa000, 0xc82000de01, 0x4, 0x0)
        <autogenerated>:1 +0x6f
github.com/mholt/caddy/server.(*Server).ServeHTTP(0xc82016c070, 0xbd1200, 0xc8200cadd0, 0xc8201aa000)
        /home/tw/golib/src/github.com/mholt/caddy/server/server.go:341 +0x260
net/http.serverHandler.ServeHTTP(0xc820182200, 0xbd1200, 0xc8200cadd0, 0xc8201aa000)
        /home/tw/goroot/src/net/http/server.go:2117 +0x73
net/http.(*conn).serve(0xc82008e580)
        /home/tw/goroot/src/net/http/server.go:1502 +0x491
created by net/http.(*Server).Serve
        /home/tw/goroot/src/net/http/server.go:2173 +0x351
```

**Comment by nemothekid** [comment]
Has the goroutine on [reverseproxy:234](https://github.com/mholt/caddy/blob/master/middleware/proxy/reverseproxy.go#L234), ended? I assume because you ctrl+c the client, that Copy won't end until the underlying connection closes (via timeout). Because that connection isn't closed, the backendConn never closes and that goroutine is still active.

**Comment by weingart** [comment]
You have a race on "defer c.Close()" in the client.

**Comment by tw4452852** [comment]
@nemothekid Yes, it is, the goroutine never ended.

**Comment by nemothekid** [comment]
Weird, I haven't tested it myself, but I assumed that a close on both sides of the Copy should interrupt the copy.

Looking at the latest [reverseproxy.go](https://golang.org/src/net/http/httputil/reverseproxy.go#L149), it looks like the official code, that the code is based off of, is now using a `CloseNotifier` to detect when the client goes away.

**Comment by nemothekid** [comment]
Or try uncommenting these lines 

https://github.com/mholt/caddy/blob/0890e330e2c07a1661ff6c4a7a08cdea478a2081/server/server.go#L84
```go
s := &Server{
		Server: &http.Server{
			Addr:      addr,
			TLSConfig: new(tls.Config),
			// TODO: Make these values configurable?
			 ReadTimeout:    30 * time.Second,
			 WriteTimeout:    30 * time.Second,
			// MaxHeaderBytes: 1 << 16,
		},
		tls:         useTLS,
		OnDemandTLS: useOnDemandTLS,
		vhosts:      make(map[string]virtualHost),
		startChan:   make(chan struct{}),
		connTimeout: gracefulTimeout,
	}
```

And see if the goroutine ends after 30 seconds.

**Comment by tw4452852** [comment]
@nemothekid Ah, I know the reason. Because the server never write any data any more, so the goroutine waits forever. Once there are some data to forward to client, it will return error in `io.Copy` and the goroutine will exit.

So I think this is a invalid case and a standard websocket server will ping/pong periodically.

**Comment by mholt** [comment]
In that case, I will close the pull request. Thank you for considering these things and looking into it! :+1:

**Pull request closed without merging**